torch/utils/data/sampler.py

```diff
@@ -210,3 +210,76 @@ def __len__(self):
             return len(self.sampler) // self.batch_size
         else:
             return (len(self.sampler) + self.batch_size - 1) // self.batch_size
+
+class DistributedChunkSampler(Sampler):
+    r"""This sampler introduces the distributed sampling.
+
+    Python DataLoader uses multi-processing instead of
+    multi-threading for parallelism, therefore, each worker
+    is a separate process with an identical copy of sampler.
+    Because of that, on a distributed environment, different processes
+    could read the same portion of the dataset. This sampler uses strides,
+    based on processes `rank`s, to coordinate parallel reading of chunks of data.
+
+    Each instance of `DistributedChunkSampler` must be configured with different a `rank`
+    (aka strides), so that chunk sampling happens in a round-robin fashion on each worker:
+    rank0, rank1, ..., rankN, rank0, rank1, ..., rankN, ...
+
+    For example, assume 2 workers reading the same `ChunkDataset` dataset.
+    Each worker process needs to configure their `DistributedChunkSampler`
+    so that one of them reads all even batches (e.g. rank 0)
+    while the other process reads all odd batches (e.g. rank 1).
+
+    Similarly to `DataLoader`, either a custom :attr:`sampler` can be specified or
+    the number of chunks (:attr:`num_chunks`) and :attr:`shuffle` flag.
+    In the latter case, `SequentialSampler` is used when :attr:`shuffle` is ``False``
+    and `RandomSampler` otherwise.
+
+    Args:
+        num_replicas (int): Number of workers participating in the sampling.
+        rank (int, optional): Current rank for the sampler, starting at 0.
+            Typically set during worker initialization function on distributed setup.
+        num_chunks (int, optional): Number of chunks participating in the sampling.
+            (default: ``0``)
+        shuffle (bool, optional): set to ``True`` to have the chunk indices reshuffled.
+            (default: ``False``)
+    """
+
+    def __init__(self, num_replicas, rank=0, num_chunks=0, shuffle=False):
+        super(DistributedChunkSampler, self).__init__(None)
+        assert rank >= 0, 'rank must be >= 0'
+        assert num_replicas >= 0, 'num_replicas must be >= 0'
+        assert rank <= num_replicas, 'rank must be <= num_replicas'
+        assert num_chunks >= num_replicas, 'num_chunks must be >= num_replicas'
+        self.rank = rank
+        self.num_replicas = num_replicas
+        self.num_chunks = num_chunks
+        self.shuffle = shuffle
+
+        if shuffle:
+            sampler = RandomSampler(data_source=range(num_chunks))
```

Comment:
No, they are similar but with important and subtle differences.

I pushed a revision with both samplers (modified DistributedSampler and a new DistributedChunkSampler) and we discussed in person the DistributedSampler issue, which was having all constructors arguments to be optional. For that reason, the user can instantiate the sampler in an invalid state right from beginning. Because objects should always be instantiated in a valid state, you agreed to drop  that approach :). You can still see the code changes for DistributedSampler at https://github.com/thiagocrepaldi/pytorch/pull/6/commits/7f26909375eb1c4d0ffce594893baa521374539e and we can discuss further
Additionally to the constructor issue, the source code would almost double just to add lots of if/else to prevent indices padding and to use the new parameter `num_indices` instead of the `dataset` one. Documentation would have to be carefully rewritten to explain the duality of the sampler too.